File: dcp-solutions/2024-12-08/solution.js
```javascript
class TreeNode {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function isValidBST(root) {
    return helper(null, root, Infinity, -Infinity);
}

function helper(low, root, high, val) {
    if (!root) {
        return true;
    }

    if ((low !== null && root.val <= low) || (high !== null && root.val >= high)) {
        return false;
    }

    return helper(val < root.val ? low : root.val, root.left, high, root.val)
           && helper(val > root.val ? root.val : high, root.right, val, root.val);
}
```

File: dcp-solutions/2024-12-08/solution.rb
```ruby
class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def validBST(root)
  return true if root.nil?

  min_val = Float::INFINITY
  max_val = -Float::INFINITY

  def is_valid(node, min_val, max_val)
    return false unless node

    val = node.val

    return val < max_val && val > min_val if [min_val, max_val].include?(val)

    left_valid = is_valid(node.left, min_val, val) && node.val < max_val
    right_valid = is_valid(node.right, val, max_val) && node.val > min_val

    return left_valid || right_valid
  end

  is_valid(root, min_val, max_val)
end
```